core/src/test/java/org/elasticsearch/action/support/master/TransportMasterNodeActionTests.java

```diff
@@ -0,0 +1,336 @@
+    private DiscoveryNode remoteNode;
+
+    @BeforeClass
+    public static void beforeClass() {
+        threadPool = new ThreadPool("TransportMasterNodeActionTests");
+    }
+
+    @Override
+    @Before
+    public void setUp() throws Exception {
+        super.setUp();
+        transport = new CapturingTransport();
+        clusterService = new TestClusterService(threadPool);
+        transportService = new TransportService(transport, threadPool);
+        transportService.start();
+        localNode = new DiscoveryNode("local_node", DummyTransportAddress.INSTANCE, Version.CURRENT);
+        remoteNode = new DiscoveryNode("remote_node", DummyTransportAddress.INSTANCE, Version.CURRENT);
+    }
+
+    @AfterClass
+    public static void afterClass() {
+        ThreadPool.terminate(threadPool, 30, TimeUnit.SECONDS);
+        threadPool = null;
+    }
+
+    <T> void assertListenerThrows(String msg, PlainActionFuture<T> listener, Class<?> klass) throws InterruptedException {
+        try {
+            listener.get();
+            fail(msg);
+        } catch (ExecutionException ex) {
+            assertThat(ex.getCause(), instanceOf(klass));
+        }
+    }
+
+    public static class Request extends MasterNodeRequest<Request> {
+        @Override
+        public ActionRequestValidationException validate() {
+            return null;
+        }
+    }
+
+    class Response extends ActionResponse {}
+
+    class Action extends TransportMasterNodeAction<Request, Response> {
+        Action(Settings settings, String actionName, TransportService transportService, ClusterService clusterService, ThreadPool threadPool) {
+            super(settings, actionName, transportService, clusterService, threadPool,
+                    new ActionFilters(new HashSet<>()), new IndexNameExpressionResolver(Settings.EMPTY), Request::new);
+        }
+
+        @Override
+        protected String executor() {
+            // very lightweight operation in memory, no need to fork to a thread
+            return ThreadPool.Names.SAME;
+        }
+
+        @Override
+        protected Response newResponse() {
+            return new Response();
+        }
+
+        @Override
+        protected void masterOperation(Request request, ClusterState state, ActionListener<Response> listener) throws Exception {
+            listener.onResponse(new Response()); // default implementation, overridden in specific tests
+        }
+
+        @Override
+        protected ClusterBlockException checkBlock(Request request, ClusterState state) {
+            return null; // default implementation, overridden in specific tests
+        }
+    }
+
+    public enum MasterOptions {
+        MASTER_EXISTS,
+        MASTER_DOES_NOT_EXIST,
+    }
+
+    public enum LocalNodeMasterOptions {
+        LOCAL_NODE_IS_MASTER,
+        LOCAL_NODE_IS_NOT_MASTER,
+    }
+
+    private ClusterState mockedClusterState(MasterOptions existsMaster, LocalNodeMasterOptions localNodeIsMaster) {
```

Comment:
can we add that to ClusterStateCreationUtils? It might be useful for others as well